File: php/meetup/Meetup.php
<?php

declare(strict_types=1);

function meetup_day(int $year, int $month, string $which, string $weekday): DateTimeImmutable
{
    if (strcmp($which, "teenth") == 0)
    {
        return teenth_day($year, $month, $weekday);
    }

    if (strcmp($which, "last") == 0)
    {
        return last_day($year, $month, $weekday);
    }

    return calc_day($year, $month, $which, $weekday);
}

function teenth_day(int $year, int $month, string $weekday): DateTimeImmutable
{
    $weekdays = array("Sunday", "Monday", "Tuesday", "Wednesday",
        "Thursday", "Friday", "Saturday");

    $res = new DateTimeImmutable("1972-01-01");
    $res = $res -> setDate($year, $month, 13);

    $day_of_the_week = array_search($weekday, $weekdays);

    $interval = DateInterval::createFromDateString("1 day");

    while (date("w", $res -> getTimeStamp()) != $day_of_the_week) {
        $res = $res -> add($interval);
    }

    return $res;
}

function last_day(int $year, int $month, string $weekday): DateTimeImmutable
{
    $weekdays = array("Sunday", "Monday", "Tuesday", "Wednesday",
        "Thursday", "Friday", "Saturday");

    $res = new DateTimeImmutable("1972-01-01");
    $year = ($month == 12) ? $year + 1 : $year;
    $month = ($month == 12) ? 1 : $month + 1;
    $res = $res -> setDate($year, $month, 1);

    $day_of_the_week = array_search($weekday, $weekdays);

    $interval = DateInterval::createFromDateString("1 day");
    $res = $res -> sub($interval);

    while (date("w", $res -> getTimeStamp()) != $day_of_the_week) {
        $res = $res -> sub($interval);
    }

    return $res;
}

function calc_day(int $year, int $month, string $which, string $weekday): DateTimeImmutable
{
    $weekdays = array("Sunday", "Monday", "Tuesday", "Wednesday",
        "Thursday", "Friday", "Saturday");

    $res = new DateTimeImmutable("1972-01-01");
    $res = $res -> setDate($year, $month, 1);

    $day_of_the_week = array_search($weekday, $weekdays);

    $interval = DateInterval::createFromDateString("1 day");

    while (date("w", $res -> getTimeStamp()) != $day_of_the_week) {
        $res = $res -> add($interval);
    }

    $schedule = array("first", "second", "third", "fourth", "teenth", "last");
    $days_to_add = array_search($which, $schedule) * 7;

    $res = $res -> add(DateInterval::createFromDateString("{$days_to_add} days"));

    return $res;
}

File: erlang/meetup/src/meetup.erl
-module(meetup).

-export([meetup/4]).


meetup(Year, Month, DayOfWeek, teenth) -> 
    calc_teenth(calendar:date_to_gregorian_days(Year, Month, 13), day_of_week_to_daynum(DayOfWeek), calendar:day_of_the_week(Year, Month, 13));
meetup(Year, Month, DayOfWeek, last) -> calc_last(calc_last_day(Year, Month), day_of_week_to_daynum(DayOfWeek), calendar:day_of_the_week(calc_last_day(Year, Month)));
meetup(Year, Month, _DayOfWeek, _) -> {Year, Month, 1}.


calc_teenth(GregorianDays, X, X) -> calendar:gregorian_days_to_date(GregorianDays);
calc_teenth(GregorianDays, DayOfWeek, _) -> 
        calc_teenth(GregorianDays + 1, DayOfWeek, calendar:day_of_the_week(calendar:gregorian_days_to_date(GregorianDays + 1))).


calc_last({Year, Month, Day}, X, X) -> {Year, Month, Day};
calc_last({Year, Month, Day}, DayOfWeek, _) -> calc_last({Year, Month, Day - 1}, DayOfWeek, calendar:day_of_the_week(Year, Month, Day - 1)).


day_of_week_to_daynum(monday) -> 1;
day_of_week_to_daynum(tuesday) -> 2;
day_of_week_to_daynum(wednesday) -> 3;
day_of_week_to_daynum(thursday) -> 4;
day_of_week_to_daynum(friday) -> 5;
day_of_week_to_daynum(saturday) -> 6;
day_of_week_to_daynum(sunday) -> 7.

calc_last_day(Year, Month) -> {Year, Month, calendar:last_day_of_the_month(Year, Month)}.
